src/meta/src/barrier/recovery.rs

```diff
@@ -542,69 +544,73 @@ impl GlobalBarrierManagerContext {
     ) -> MetaResult<InflightActorInfo> {
         let mgr = self.metadata_manager.as_v2_ref();
 
-        let all_inuse_parallel_units: HashSet<_> = mgr
+        // all worker slots used by actors
+        let all_inuse_worker_slots: HashSet<_> = mgr
             .catalog_controller
-            .all_inuse_parallel_units()
+            .all_inuse_worker_slots()
             .await?
             .into_iter()
             .collect();
 
-        let active_parallel_units: HashSet<_> = active_nodes
+        let active_worker_slots: HashSet<_> = active_nodes
             .current()
             .values()
-            .flat_map(|node| node.parallel_units.iter().map(|pu| pu.id as i32))
+            .flat_map(|node| {
+                (0..node.parallelism).map(|idx| WorkerSlotId::new(node.id, idx as usize))
+            })
             .collect();
 
-        let expired_parallel_units: BTreeSet<_> = all_inuse_parallel_units
-            .difference(&active_parallel_units)
+        let expired_worker_slots: BTreeSet<_> = all_inuse_worker_slots
+            .difference(&active_worker_slots)
             .cloned()
             .collect();
-        if expired_parallel_units.is_empty() {
+
+        if expired_worker_slots.is_empty() {
             debug!("no expired parallel units, skipping.");
             return self.resolve_actor_info(active_nodes).await;
         }
 
         debug!("start migrate actors.");
-        let mut to_migrate_parallel_units = expired_parallel_units.into_iter().rev().collect_vec();
+        let mut to_migrate_worker_slots = expired_worker_slots.into_iter().rev().collect_vec();
         debug!(
             "got to migrate parallel units {:#?}",
-            to_migrate_parallel_units
+            to_migrate_worker_slots
         );
-        let mut inuse_parallel_units: HashSet<_> = all_inuse_parallel_units
-            .intersection(&active_parallel_units)
+
+        let mut inuse_worker_slots: HashSet<_> = all_inuse_worker_slots
+            .intersection(&active_worker_slots)
             .cloned()
             .collect();
 
         let start = Instant::now();
         let mut plan = HashMap::new();
-        'discovery: while !to_migrate_parallel_units.is_empty() {
-            let new_parallel_units = active_nodes
+        'discovery: while !to_migrate_worker_slots.is_empty() {
+            let new_worker_slots = active_nodes
                 .current()
                 .values()
                 .flat_map(|node| {
-                    node.parallel_units
-                        .iter()
-                        .filter(|pu| !inuse_parallel_units.contains(&(pu.id as _)))
+                    (0..node.parallelism)
```

Comment:
Using `parallelism` as the number of initial worker slots can be problematic. When the number of actors on CN exceeds `parallelism` and a reduction in configuration triggers migration, the cluster will not recover due to insufficient slots.